### PR TITLE
fix: Replace cloneDeep with klona #12936

### LINKED ISSUE #12936: [Task]: Replace cloneDeep with klona
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Description

As `klona` is faster when compared to lodash's cloneDeep. We would like to migrate from `cloneDeep` to `klona` in all the places where it is possible in client code. 

### Steps 
- [ ] Search for `cloneDeep` in client code
- [ ] Replace `cloneDeep` with `klona` 
- [ ] Add tests to verify you changes works as expected before and after the change.

### PR DESCRIPTION
**Pull request by JaiZemoso**
## Description

Fixes #12936 
_or_  
Issue URL  `[Issue URL]( https://github.com/appsmithorg/appsmith/issues/12936)`
> [!WARNING]  
> _If no issue exists, please create an issue first, and check with the maintainers if the issue is valid._

## Automation

/ok-to-test tags=""

### :mag: Cypress test results

> [!CAUTION]  
> If you modify the content in this section, you are likely to disrupt the CI result for your PR.

## Communication
Should the DevRel and Marketing teams inform users about this change?
- [ ] Yes
- [ ] No

## Summary by CodeRabbit

- **Refactor**
  - Improved the performance and efficiency of the application by replacing `lodash.cloneDeep` with `klona` for deep cloning objects across various components and utilities.

These changes enhance object copying mechanisms, potentially reducing memory usage and improving application speed without altering the existing functionality from the end-user perspective.

### CODE REVIEW

File: app/client/src/widgets/ListWidget/widget/index.tsx [hunk 1]
```diff
@@ -45,6 +44,7 @@ import {
   toString,
```
[the remark this answers is not in the document]
Comment (JaiZemoso) [reply app/client/src/widgets/ListWidget/widget/index.tsx:44]: done

File: app/client/src/widgets/ListWidget/widget/index.tsx [hunk 2]
```diff
@@ -45,6 +44,7 @@ import {
...
   xor,
 } from "lodash";
+import { klona } from "klona";
```
[the remark this answers is not in the document]
Comment (JaiZemoso) [reply app/client/src/widgets/ListWidget/widget/index.tsx:47]: done

File: app/client/src/sagas/WidgetOperationSagas.tsx [hunk 3]
```diff
@@ -32,7 +32,8 @@ import {
   RenderModes,
   WIDGET_ID_SHOW_WALKTHROUGH,
 } from "constants/WidgetConstants";
-import _, { cloneDeep, get, isString, set, uniq } from "lodash";
+import _, { get, isString, set, uniq } from "lodash";
+import { klona } from "klona";
```
[the remark this answers is not in the document]
Comment (JaiZemoso) [reply app/client/src/sagas/WidgetOperationSagas.tsx:36]: done

**Comment by JaiZemoso** [comment]
@rishabhrathod01 can you kindly review this PR ?

**Comment by rishabhrathod01** [comment]
@JaiZemoso the task is not as simple as the title. We cannot directly replace the library here. Each such replacement will need to be tested either manually or using automation. As this would be quite a huge change hence this task was removed from the `Good first issue`.

**Comment by rishabhrathod01** [comment]
@somangshu code where the library is replaced is majorly widgets related. Could you please assign a reviewer from widget team for this PR.

**Comment by somangshu** [comment]
@rishabhrathod01 this is a very large change and hence risky, its not even a GFI. I suggest we dont take it forward. 

cc @rajatagrawal what do you think so?

**Comment by akshayvijayjain** [comment]
@somangshu @rishabhrathod01 @rajatagrawal 

Please excuse me jumping in between.

If you see the changes is only one

replace cloneDeep with klona.

both the functions do same thing i.e. create deep clone.

if I ask chatgpt, it also says klona is a drop in replacement for cloneDeep, and we did it over here.

if klona is better and it is a drop-in replacement for cloneDeep, what risk do you see?

**Comment by somangshu** [comment]
ChatGPT does not understand the context of appsmith repo. It does not know the impact that the library change might have. 

AFAIK, we have tried this before and did not succeed at completing it. There were test failures noticed. 
Ill let resident technical experts in our team take the final call.

**Comment by rishabhrathod01** [comment]
> @rishabhrathod01 this is a very large change and hence risky, its not even a GFI. I suggest we dont take it forward. 
> 
> 
> 
> cc @rajatagrawal what do you think?

Yes, i agree. The result from klona and clone of lodash has differences for date objects, regex values and more that we discovered in past. We need to be careful before making such change.

**Comment by rishabhrathod01** [comment]
> @rishabhrathod01 this is a very large change and hence risky, its not even a GFI. I suggest we dont take it forward. 
> 
> 
> 
> cc @rajatagrawal what do you think?

Yes, i agree. The result from klona and clone of lodash has differences for date objects, regex values and more that we discovered in past. We need to be careful before making such change.

**Pull request closed without merging**